File: UI/src/app/core/utils/distance.ts
import { Customer, Depot, RoutePlan } from '../models';

export function euclideanDistance(ax: number, ay: number, bx: number, by: number): number {
  return Math.hypot(ax - bx, ay - by);
}

export function distanceBetweenNodes(
  aId: number,
  bId: number,
  depot: Depot,
  customers: Customer[],
): number {
  if (aId === 0) {
    const b = customers.find((customer) => customer.id === bId);
    return b ? euclideanDistance(depot.x, depot.y, b.x, b.y) : 0;
  }

  if (bId === 0) {
    const a = customers.find((customer) => customer.id === aId);
    return a ? euclideanDistance(a.x, a.y, depot.x, depot.y) : 0;
  }

  const a = customers.find((customer) => customer.id === aId);
  const b = customers.find((customer) => customer.id === bId);
  if (!a || !b) {
    return 0;
  }
  return euclideanDistance(a.x, a.y, b.x, b.y);
}

export function computeRouteDistance(route: RoutePlan, depot: Depot, customers: Customer[]): number {
  let distance = 0;
  const nodes = route.nodes;
  for (let i = 0; i < nodes.length - 1; i += 1) {
    distance += distanceBetweenNodes(nodes[i], nodes[i + 1], depot, customers);
  }
  return Number(distance.toFixed(2));
}

File: backend/src/main/java/com/cvrp/util/Distance.java
package com.cvrp.util;

import com.cvrp.model.Customer;
import com.cvrp.model.Instance;

import java.util.List;

public final class Distance {
    private Distance() {
    }

    public static double[][] buildMatrix(Instance instance) {
        List<Customer> customers = instance.customers();
        int size = customers.size() + 1;
        double[][] matrix = new double[size][size];
        for (int i = 0; i < size; i++) {
            for (int j = i; j < size; j++) {
                double dist = compute(instance, customers, i, j);
                matrix[i][j] = dist;
                matrix[j][i] = dist;
            }
        }
        return matrix;
    }

    private static double compute(Instance instance, List<Customer> customers, int from, int to) {
        double x1;
        double y1;
        if (from == 0) {
            x1 = instance.depot().x();
            y1 = instance.depot().y();
        } else {
            Customer c = customers.get(from - 1);
            x1 = c.x();
            y1 = c.y();
        }

        double x2;
        double y2;
        if (to == 0) {
            x2 = instance.depot().x();
            y2 = instance.depot().y();
        } else {
            Customer c = customers.get(to - 1);
            x2 = c.x();
            y2 = c.y();
        }
        double dx = x1 - x2;
        double dy = y1 - y2;
        return Math.hypot(dx, dy);
    }
}
